[Beck] Step2 - Create Card class & Summarize Step1

SangHwi-Back opened the pull request:
## 학습 내용

* 객체 지향 프로그래밍을 위해 Card 클래스는 행동만을 정의하고 ```Card 클래스```의 심볼과 숫자는 ```CardInfo 구조체```를 이용하여 작성하였습니다.
* ```typealias``` 키워드를 이용하고 해당 타입을 확장해보면서, 하나의 타입을 다양한 방식으로 사용할 수 있다는 점을 학습했습니다.
* ```팩토리 디자인 패턴```을 실습해 보았습니다. Card 클래스 하나를 정의하는 것은 Card 클래스 안에 그대로 두었지만, 무작위적으로 카드를 생성하거나, 한 덱의 카드묶음을 생성하는 것은 팩토리 클래스를 이용할 수 있도록 하였습니다.
* 이전 스텝에서 사용한 ```UIImage``` ```UIImageView``` ```UIStackView``` 클래스에 대한 애플 공식문서를 보고 내용을 간단히 정리하였습니다.

### 아쉬운 점

참고자료로 제공된 두 개의 자료인 ```Swift 타입 별 메모리 분석```, ```스택과 힙 - 프로세스 메모리``` 강의를 완벽히 이해하고 코드를 작성하지는 못했습니다.

작성한 코드와 자료를 반복적으로 번갈아 보면서 이해하면서 코드를 작성했다면 좀 더 질 높은 코드 작성이 가능하지 않았을까 하는 아쉬움이 남습니다.

README에 작업 종료시간을 기재해 놓았지만 이는 추측되는 시간입니다. 따로 메모를 해두지 않고 Rebase를 진행해버렸습니다. 이번에는 커밋 메시지에 년월일로 기록해 놓는 것으로 실수를 방지하였습니다.

### 고민되었던 부분

1. CardInfo를 구조체로 할지 클래스로 할지 잠시 고민해 보았습니다.   
하지만 개인적으로는 클래스에서 데이터 역할을 하는 것이고, 재사용 될 가능성도 있어서(예: 하나의 카드를 만든 뒤 같은 심볼의 카드를 여러 개 만들 경우 CardInfo만 복사) 구조체로 만들었습니다.
가장 우려되는 부분은 순환참조였는데 이는 참고자료를 다시 한번 상세히 읽어보고 판단해봐야할 것 같습니다.

2. 팩토리 디자인 패턴을 사용하는게 맞는지도 고민해 보았습니다.   
팩토리 디자인 패턴을 사용하게 되면서 static 함수를 생성했는데, static 요소들은 객체화에 상관없이 메모리에 적재되는 것으로 알고 있습니다. 앱의 성능에 부담을 주지 않을까 생각을 해 보았습니다.
확실히 코드를 명확하게 하는데에는 도움을 받은 것 같습니다.

SangHwi-Back committed:
Modify card create logic change

카드 생성 로직을 스택뷰를 이용하여 보여줄 수 있도록 소스 변경하였습니다.

SangHwi-Back committed:
Make Card class as object-oriented

객체지향 연습을 위해 Card 클래스를 만드는 로직을 만들었습니다.
특히 구조에 신경을 많이 쓰게 되었습니다.
Card 클래스는 Card의 행동, CardInfo는 Card의 데이터를 담당하도록 하였습니다.
extension 문법도 웬만하면 따로 파일을 구분하여 작성하였습니다.
extension 문법은 대부분 primitive type들을 확장하는 경우가 많아서 내용이 결국 길어지는 걸 경험하고 나서 생긴 버릇입니다.
시작시간: 2022/2/22 19:30
종료시간: 2022/2/22 23:00

SangHwi-Back committed:
Update README about step1

Step1을 진행한 방식과 관련 내용들을 업데이트 합니다.
Step2에 대한 내용들도 기존 방식과 같이 Rebase 이후 진행할 예정입니다.

SangHwi-Back committed:
Update gitignore for AppCode

AppCode에서 프로젝트를 열면 생성되는 파일에 대한 gitignore 설정을 업데이트 합니다.

godrm commented on PokerGameApp/PokerGameApp/Card+Array+Extension.swift:
```diff
@@ -0,0 +1,20 @@
+    var readAll: String {
+        self.reduce("") { partialResult, card in
+            partialResult + " " + card.read
+        }
```
이미 Array는 description으로 이런 기능을 제공합니다. 한 번 확인해보세요

godrm commented on PokerGameApp/PokerGameApp/Card+Enum.swift:
```diff
@@ -0,0 +1,20 @@
+    case spade = "♠️"
+    case clover = "♣️"
+    case heart = "♥️"
+    case diamond = "♦️"
```
이런 하위 타입은 Card.Symbol 형태로 nested로 선언하는 것을 고려해보세요

godrm commented on PokerGameApp/PokerGameApp/CardFactory.swift:
```diff
@@ -0,0 +1,34 @@
+        }
+        
+        return result
+    }
```
오 팩토리가 있다는 것은 좋은 접근입니다. 
팩토리가 있으면 어떤 장점이 있을까요?

godrm commented on PokerGameApp/PokerGameApp/Card.swift:
```diff
@@ -0,0 +1,31 @@
+    
+    var read: String {
+        "\(info.symbol.rawValue)\(info.number.encodedNum)"
+    }
```
스위프트 객체들은 문자열을 제공하기 위해서 description을 활용합니다. CustomStringConvertible 프로토콜을 확인해보세요

godrm commented on PokerGameApp/PokerGameApp/Card.swift:
```diff
@@ -0,0 +1,31 @@
+    }
+    
+    func isMatching(with cardInfo: CardInfo) -> Bool {
+        return self.info == cardInfo
```
Card와 CardInfo가 합쳐져도 되지 않을까요?
따로 구분한 이유와 기준은 무엇인지 궁금합니다

godrm commented on PokerGameApp/PokerGameApp/CardInfo.swift:
```diff
@@ -0,0 +1,40 @@
+    // 초기화 시 지정하고 나서 바꿀 일은 없지만 참조할 필요는 있을 것 같아서
+    // 접근제한자를 private(set)으로 제한하였습니다.
+    private(set) var number: CardNumber!
+    private(set) var symbol: CardSymbol!
```
그럼 let으로 선언하는 것과 어떤 차이가 있을까요

godrm commented on PokerGameApp/PokerGameApp/CardNumber+Extension.swift:
```diff
@@ -0,0 +1,44 @@
+
+// 해당 extension은 Int를 CardNumber로 바꾸기 쉽도록 만들거나
+// CardNumber를 실제 카드에 인쇄되는 문자로 바꾸어줍니다.
+extension CardNumber {
```
CardNumber를 Int로 했는데 값 범위가 핵심인 것 같습니다
결국 Int라서 CardNumber를 초기화할 때 0, 1234 이런 카드가 생길수도 있지 않을까요?

godrm commented on PokerGameApp/PokerGameApp/String+Extension.swift:
```diff
@@ -0,0 +1,15 @@
+    var convertCardNumber: CardNumber? {
+        guard let number = Int(self) else { return nil }
+        return CardNumber.allRange.contains(number as CardNumber) ? number as CardNumber : nil
+    }
```
위에 피드백과 이어지는 데 만약에 Int가 아니라 타입을 지정했다면 이 코드는 불필요할 수 있습니다

godrm reviewed (changes requested):
다음 피드백을 확인해보시고 Card부분을 수정해서 Push해보세요

SangHwi-Back committed:
Rewrite code following comment from PR

CardInfo의 Number와 Symbol을 Card의 Nested Type으로 변경합니다.
두 속성은 트럼프 카드에서만 쓰이기 때문입니다.
특히 Card.Number의 범위는 CardInfo를 초기화할 때 검사합니다.
Card의 Equatable과 CardInfo의 Equatable을 나눕니다.
변수의 값을 구분하는 것에 따로 비교 메소드를 가질 필요는 없었습니다.
이를 통해 불필요한 코드를 제거하였습니다.
